primary/src/block_synchronizer/mock.rs

```diff
@@ -0,0 +1,170 @@
+    rx_core: Receiver<Core<PublicKey>>,
+}
+
+impl<PublicKey: VerifyingKey> MockBlockSynchronizerCore<PublicKey> {
+    async fn run(&mut self) {
+        loop {
+            tokio::select! {
+                Some(command) = self.rx_commands.recv() => {
+                    match command {
+                        Command::SynchronizeBlockHeaders { block_ids, respond_to } => {
+                            let (times, results) = self
+                                .block_headers_expected_requests
+                                .remove(&block_ids)
+                                .unwrap_or_else(||panic!("{}", format!("Unexpected call received for SynchronizeBlockHeaders, {:?}", block_ids).as_str()))
+                                .to_owned();
+
+                            if times > 1 {
+                                self.block_headers_expected_requests.insert(block_ids, (times - 1, results.clone()));
+                            }
+
+                            for result in results {
+                                respond_to.send(result).await.expect("Couldn't send message");
+                            }
+                        }
+                        Command::SynchronizeBlockPayload { .. } => {}
+                    }
+                }
+                Some(command) = self.rx_core.recv() => {
+                    match command {
+                        Core::SynchronizeBlockHeaders {
+                            block_ids,
+                            times,
+                            result,
+                            ready,
+                        } => {
+                            self.block_headers_expected_requests.insert(block_ids, (times, result));
+                            ready.send(()).expect("Failed to send ready message");
+                        },
+                        Core::AssertExpectations {ready} => {
+                            self.assert_expectations();
+                            ready.send(()).expect("Failed to send ready message");
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    fn assert_expectations(&self) {
+        let mut result: String = "".to_string();
+
+        for (ids, results) in &self.block_headers_expected_requests {
+            result.push_str(
+                format!(
+                    "SynchronizeBlockHeaders, ids={:?}, results={:?}",
+                    ids, results
+                )
+                .as_str(),
+            );
+        }
+
+        if !result.is_empty() {
+            panic!(
+                "There are expected calls that haven't been fulfilled \n\n {}",
+                result
+            );
+        }
+    }
+}
+
+/// A mock helper for the BlockSynchronizer to help us mock the responses
+/// eliminating the need to wire in the actual BlockSynchronizer when needed
+/// for other components.
+pub struct MockBlockSynchronizer<PublicKey: VerifyingKey> {
```

Comment:
Unfortunately I had to do the separation between the `MockBlockSynchronizer` and the "inner" struct `MockBlockSynchronizerCore` to allow me have a way to interface with the engine after we have spin up the task. Otherwise I wouldn't be able to perform follow up calls, like the `assert_expectations` since my object would be moved. That complicated things a bit (thanks Rust) but I consider it still straightforward